src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -0,0 +1,167 @@
+object ReadStructure {
+  /** Creates a read structure from a string */
+  def apply(readStructure: String): ReadStructure = {
+    var offset = 0
+    var startDigit = 0
+    val rs = new ReadStructure(readStructure.zipWithIndex.flatMap { case (c, idx) =>
+      if (!c.isDigit) {
+        val endDigit  = idx
+        if (endDigit <= startDigit) throw new IllegalArgumentException("Read structure missing a segment length: " + formatReadStructureError(readStructure, idx, idx+1))
+        val length    = Integer.parseInt(readStructure.substring(startDigit, endDigit))
+        val curOffset = offset
+        if (length == 0) throw new IllegalArgumentException("Read structure had a segment with zero length: " + formatReadStructureError(readStructure, startDigit, endDigit))
+        startDigit = idx + 1
+        offset    += length
+        try {
+          Some(ReadSegment(offset = curOffset, length = length, c = c))
+        } catch {
+          case ex: IllegalArgumentException =>
+            throw new IllegalArgumentException("Read structure had a segment of unknown type: " + formatReadStructureError(readStructure, idx, idx+1))
+        }
+      }
+      else {
+        None
+      }
+    })
+    if (startDigit < readStructure.length) throw new IllegalArgumentException("Read structure had trailing digits: " + formatReadStructureError(readStructure, startDigit, readStructure.length))
+    rs
+  }
+
+  /**
+    * Inserts square brackets around the characers in the read structure that are causing the error.
+    * @param s the read structure string
+    * @param start the start of the error in the string (inclusive)
+    * @param end the ned of the error in the string (exclusive)
+    */
+  private def formatReadStructureError(s: String, start: Int, end: Int) = {
+    val prefix = s.substring(0, start)
+    val error  = s.substring(start, end)
+    val suffix = if (end == s.length) "" else s.substring(end, s.length)
+    s"$prefix[$error]$suffix"
+  }
+}
+
+/**
+  * Describes the structure of a give read.  A read contains one or more read segments. A read segment describes
+  * a contiguous stretch of bases of the same type (ex. template bases) of some length and some offset from the start
+  * of the read.
+  */
+class ReadStructure(segs: Seq[ReadSegment]) extends immutable.Seq[ReadSegment] {
+  private val segments: Seq[ReadSegment] = segs
+  def length: Int = segments.length
+  def apply(idx: Int): ReadSegment = segments(idx)
+  def iterator: Iterator[ReadSegment] = segments.iterator
+  override def toString: String = segments.map(_.toString).mkString
+  def totalBases: Int = segments.map(_.length).sum
+
+  /** Splits the given bases into tuples with its associated read segment. */
+  def extract(bases: String, strict: Boolean = true): Seq[(String, ReadSegment)] = {
```

Comment:
Coming at this fresh it took me a minute to realize this is breaking apart an actual read.  Perhaps calling it something like `structureRead` would be clearer?  A small example in the scaladoc would also be super helpful.